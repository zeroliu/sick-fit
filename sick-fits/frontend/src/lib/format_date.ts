export function formatDate(dateStr: string) {
  const date = new Date(dateStr);
  const options = {
    year: 'numeric',
    month: 'numeric',
    day: 'numeric',
    hour: 'numeric',
    minute: 'numeric',
    second: 'numeric',
    timeZoneName: 'short',
  };
  return new Intl.DateTimeFormat('en-US', options).format(date);
}
