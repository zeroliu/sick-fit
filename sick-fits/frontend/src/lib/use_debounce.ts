import { useRef, useCallback, useEffect } from 'react';

export function useDebounce<Args extends unknown[]>(
  callback: (...args: Args) => void,
  timeout: number,
): (...args: Args) => void {
  const tid = useRef(-1);
  const debouncedFunc = useRef(callback);
  const hasUnmounted = useRef(false);
  debouncedFunc.current = callback;

  useEffect(
    () => () => {
      hasUnmounted.current = true;
    },
    [],
  );

  const debouncedCallback = useCallback(
    (...args: Args) => {
      clearTimeout(tid.current);
      tid.current = setTimeout(() => {
        if (hasUnmounted.current) {
          return;
        }
        debouncedFunc.current(...args);
      }, timeout);
    },
    [timeout],
  );

  return debouncedCallback;
}
