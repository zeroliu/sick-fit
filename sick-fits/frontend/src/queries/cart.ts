import { MutationHookOptions, useMutation } from '@apollo/react-hooks';
import gql from 'graphql-tag';

import {
  Mutation,
  MutationAddToCartArgs,
  MutationRemoveFromCartArgs,
} from 'src/generated/graphql';

export const ADD_TO_CART_MUTATION = gql`
  mutation($data: AddToCartInput!) {
    addToCart(data: $data)
  }
`;
interface AddToCartMutationData {
  addToCart: Mutation['addToCart'];
}
export function useAddToCartMutation(
  options?: MutationHookOptions<AddToCartMutationData, MutationAddToCartArgs>,
) {
  return useMutation(ADD_TO_CART_MUTATION, options);
}

export const REMOVE_FROM_CART_MUTATION = gql`
  mutation($data: RemoveFromCartInput!) {
    removeFromCart(data: $data)
  }
`;
interface RemoveFromCartMutationData {
  removeFromCart: Mutation['removeFromCart'];
}
export function useRemoveFromCartMutation(
  options?: MutationHookOptions<
    RemoveFromCartMutationData,
    MutationRemoveFromCartArgs
  >,
) {
  return useMutation(REMOVE_FROM_CART_MUTATION, options);
}
