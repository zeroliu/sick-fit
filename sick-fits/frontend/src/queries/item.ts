import {
  useQuery,
  QueryHookOptions,
  MutationHookOptions,
  useMutation,
} from '@apollo/react-hooks';
import gql from 'graphql-tag';

import {
  Mutation,
  Query,
  QueryItemsArgs,
  QueryItemArgs,
  MutationCreateItemArgs,
  MutationUpdateItemArgs,
  MutationDeleteItemArgs,
} from 'src/generated/graphql';

export interface Item {
  id: string;
  title: string;
  price: number;
  description: string;
  image?: string;
  largeImage?: string;
}

const ALL_ITEMS_QUERY = gql`
  query($skip: Int, $take: Int) {
    items(skip: $skip, take: $take) {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;
interface AllItemsQueryData {
  items: Item[];
}
export function useAllItemsQuery(
  options?: QueryHookOptions<AllItemsQueryData, QueryItemsArgs>,
) {
  return useQuery(ALL_ITEMS_QUERY, options);
}

const ITEM_QUERY = gql`
  query($id: ID!) {
    item(id: $id) {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;
interface ItemQueryData {
  item?: Item;
}
export function useItemQuery(
  options?: QueryHookOptions<ItemQueryData, QueryItemArgs>,
) {
  return useQuery(ITEM_QUERY, options);
}

const ITEMS_CONNECTION_QUERY = gql`
  query {
    itemsConnection {
      totalCount
    }
  }
`;
interface ItemsConnectionQueryData {
  itemsConnection: Query['itemsConnection'];
}
export function useItemsConnectionQuery() {
  return useQuery<ItemsConnectionQueryData>(ITEMS_CONNECTION_QUERY);
}

const CREATE_ITEM_MUTATION = gql`
  mutation($data: CreateItemInput!) {
    createItem(data: $data) {
      id
    }
  }
`;
interface CreateItemMutationData {
  createItem: { id: number };
}
export function useCreateItemMutation(
  options?: MutationHookOptions<CreateItemMutationData, MutationCreateItemArgs>,
) {
  return useMutation(CREATE_ITEM_MUTATION, options);
}

const UPDATE_ITEM_MUTATION = gql`
  mutation($id: ID!, $data: UpdateItemInput!) {
    updateItem(id: $id, data: $data)
  }
`;
interface UpdateItemMutationData {
  updateItem: Mutation['updateItem'];
}
export function useUpdateItemMutation(
  options?: MutationHookOptions<UpdateItemMutationData, MutationUpdateItemArgs>,
) {
  return useMutation(UPDATE_ITEM_MUTATION, options);
}

const DELETE_ITEM_MUTATION = gql`
  mutation($id: ID!) {
    deleteItem(id: $id)
  }
`;
interface DeleteItemMutationData {
  deleteItem: Mutation['deleteItem'];
}
export function useDeleteItemMutation(
  options?: MutationHookOptions<DeleteItemMutationData, MutationDeleteItemArgs>,
) {
  return useMutation(DELETE_ITEM_MUTATION, options);
}
