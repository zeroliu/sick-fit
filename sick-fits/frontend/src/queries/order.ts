import { QueryHookOptions, useMutation, useQuery } from '@apollo/react-hooks';
import gql from 'graphql-tag';

import {
  MutationPayArgs,
  OrderItem as GeneratedOrderItem,
  Order as GeneratedOrder,
  User,
  QueryOrderArgs,
} from 'src/generated/graphql';

const PAY_MUTATION = gql`
  mutation($data: PayInput!) {
    pay(data: $data) {
      id
    }
  }
`;
interface PayMutationData {
  pay: Pick<GeneratedOrder, 'id'>;
}
export function usePayMutation(
  options?: QueryHookOptions<PayMutationData, MutationPayArgs>,
) {
  return useMutation(PAY_MUTATION, options);
}

const ORDER_QUERY = gql`
  query($orderId: ID!) {
    order(orderId: $orderId) {
      id
      paymentIntent
      total
      createdAt
      user {
        id
      }
      items {
        id
        title
        description
        price
        image
        quantity
      }
    }
  }
`;
type OrderItem = Pick<
  GeneratedOrderItem,
  'id' | 'title' | 'description' | 'price' | 'image' | 'quantity'
>;
export type Order = Pick<
  GeneratedOrder,
  'id' | 'paymentIntent' | 'total' | 'createdAt'
> & {
  user: Pick<User, 'id'>;
  items: OrderItem[];
};
interface OrderQueryData {
  order: Order;
}
export function useOrderQuery(
  options?: QueryHookOptions<OrderQueryData, QueryOrderArgs>,
) {
  return useQuery(ORDER_QUERY, options);
}
