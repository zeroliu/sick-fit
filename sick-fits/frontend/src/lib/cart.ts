import { CartItem } from 'src/queries/user';

export function calcTotalPrice(cartItems: CartItem[]) {
  return cartItems.reduce((tally, cartItem) => {
    if (!cartItem.item) return tally;
    return tally + cartItem.quantity * cartItem.item.price;
  }, 0);
}

export function countItems(cartItems: CartItem[]) {
  return cartItems.reduce((tally, cartItem) => {
    if (!cartItem.item) return tally;
    return tally + cartItem.quantity;
  }, 0);
}
