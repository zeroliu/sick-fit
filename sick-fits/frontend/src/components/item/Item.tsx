import React from 'react';
import { Item as ItemType } from 'src/queries/item';
import { Item as StyledItem } from './item_styles';
import { Title } from 'src/components/styles/Title';
import Link from 'next/link';
import { PriceTag } from 'src/components/styles/PriceTag';
import { formatMoney } from 'src/lib/format_money';
import { DeleteItem } from '../delete_item/DeleteItem';

interface Props {
  data: ItemType;
}

export const Item: React.FC<Props> = ({ data }) => {
  return (
    <StyledItem>
      {data.image && <img src={data.image} alt={data.title}></img>}
      <Title>
        <Link
          href={{
            pathname: '/item',
            query: { id: data.id },
          }}>
          <a>{data.title}</a>
        </Link>
      </Title>
      <PriceTag>{formatMoney(data.price)}</PriceTag>
      <p>{data.description}</p>
      <div className='buttonList'>
        <Link
          href={{
            pathname: '/update',
            query: { id: data.id },
          }}>
          <a>Edit ✏️</a>
        </Link>
        <button>Add to Cart</button>
        <DeleteItem id={data.id}>Delete</DeleteItem>
      </div>
    </StyledItem>
  );
};
