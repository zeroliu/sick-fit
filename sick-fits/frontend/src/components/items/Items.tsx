import React from 'react';

import { Center, ItemsList } from './items_styles';
import { Item } from 'src/components/item/Item';
import { ItemsPagination } from 'src/components/items_pagination/ItemsPagination';
import { perPage } from 'src/config';
import { useAllItemsQuery } from 'src/queries/item';


interface Props {
  currentPage: number;
}

export const Items: React.FC<Props> = ({ currentPage }) => {
  const { loading, error, data } = useAllItemsQuery({
    variables: {
      take: perPage,
      skip: (currentPage - 1) * perPage,
    },
  });
  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error {error.message}</p>;
  if (!data) return <p>Error: Empty data</p>;
  return (
    <Center>
      <ItemsPagination currentPage={currentPage}></ItemsPagination>
      <ItemsList>
        {data.items.map((item) => (
          <Item key={item.id} data={item}></Item>
        ))}
      </ItemsList>
      <ItemsPagination currentPage={currentPage}></ItemsPagination>
    </Center>
  );
};
