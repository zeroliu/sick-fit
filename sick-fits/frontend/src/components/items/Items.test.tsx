import { MockedProvider, MockedResponse } from '@apollo/react-testing';
import casual from 'casual';
import { mount } from 'enzyme';
import React from 'react';

import { Items } from './Items';
import { actAfterRender } from 'src/lib/test_utils';
import { ALL_ITEMS_QUERY, ITEMS_CONNECTION_QUERY } from 'src/queries/item';

casual.seed(123);

const mocks: MockedResponse[] = [
  {
    request: {
      query: ITEMS_CONNECTION_QUERY,
    },
    result: {
      data: {
        itemsConnection: {
          __typename: 'ItemsConnection',
          totalCount: 5,
        },
      },
    },
  },
  {
    request: {
      query: ALL_ITEMS_QUERY,
      variables: { take: 4, skip: 0 },
    },
    result: {
      data: {
        items: [
          {
            __typename: 'Item',
            id: '123',
            title: casual.title,
            price: casual.integer,
            description: casual.sentence,
            image: 'foo.jpg',
            largeImage: 'large_foo.jpg',
          },
          {
            __typename: 'Item',
            id: '456',
            title: casual.title,
            price: casual.integer,
            description: casual.sentence,
            image: 'foo.jpg',
            largeImage: 'large_foo.jpg',
          },
        ],
      },
    },
  },
];

describe('<Items />', () => {
  it('renders loading state correctly', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Items currentPage={1} />
      </MockedProvider>,
    );
    expect(wrapper.find(Items)).toMatchSnapshot();
    await actAfterRender();
  });

  it('renders final state correctly', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Items currentPage={1} />
      </MockedProvider>,
    );
    await actAfterRender();
    wrapper.update();
    expect(wrapper.find(Items)).toMatchSnapshot();
  });
});
