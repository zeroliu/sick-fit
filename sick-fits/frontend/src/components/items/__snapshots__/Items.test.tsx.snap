// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Items /> renders final state correctly 1`] = `
<Items
  currentPage={1}
>
  <items_styles__Center>
    <div
      className="items_styles__Center-sc-1jzjd2i-0 eFCBDz"
    >
      <ItemsPagination
        currentPage={1}
      >
        <items_pagination_styles__StyledItemsPagination>
          <div
            className="items_pagination_styles__StyledItemsPagination-sc-1do2rc3-0 kTuLOm"
          >
            <Head>
              <Component
                handleStateChange={null}
                inAmpMode={false}
                reduceComponentsToState={[Function]}
              />
            </Head>
            <Link
              href={
                Object {
                  "pathname": "items",
                  "query": Object {
                    "page": 0,
                  },
                }
              }
            >
              <a
                aria-disabled={true}
                href="items?page=0"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                Prev
              </a>
            </Link>
            <p>
              Page 
              1
               of 
              2
            </p>
            <Link
              href={
                Object {
                  "pathname": "items",
                  "query": Object {
                    "page": 2,
                  },
                }
              }
            >
              <a
                aria-disabled={false}
                href="items?page=2"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                Next
              </a>
            </Link>
          </div>
        </items_pagination_styles__StyledItemsPagination>
      </ItemsPagination>
      <items_styles__ItemsList>
        <div
          className="items_styles__ItemsList-sc-1jzjd2i-1 gyVOSp"
        >
          <Item
            data={
              Object {
                "__typename": "Item",
                "description": "Est omnis porro porro quos.",
                "id": "123",
                "image": "foo.jpg",
                "largeImage": "large_foo.jpg",
                "price": [Function],
                "title": "Quo quis molestiae",
              }
            }
            key="123"
          >
            <item_styles__Item>
              <div
                className="item_styles__Item-sc-1wrkfgx-0 aVlSA"
              >
                <img
                  alt="Quo quis molestiae"
                  src="foo.jpg"
                />
                <Title>
                  <h3
                    className="Title-sc-1002kw6-0 eBuTHz"
                  >
                    <Link
                      href={
                        Object {
                          "pathname": "/item",
                          "query": Object {
                            "id": "123",
                          },
                        }
                      }
                    >
                      <a
                        href="/item?id=123"
                        onClick={[Function]}
                        onMouseEnter={[Function]}
                      >
                        Quo quis molestiae
                      </a>
                    </Link>
                  </h3>
                </Title>
                <PriceTag>
                  <span
                    className="PriceTag-sc-145cui2-0 cUpSNy"
                  >
                    $NaN
                  </span>
                </PriceTag>
                <p>
                  Est omnis porro porro quos.
                </p>
                <div
                  className="buttonList"
                >
                  <Link
                    href={
                      Object {
                        "pathname": "/update",
                        "query": Object {
                          "id": "123",
                        },
                      }
                    }
                  >
                    <a
                      href="/update?id=123"
                      onClick={[Function]}
                      onMouseEnter={[Function]}
                    >
                      Edit ✏️
                    </a>
                  </Link>
                  <AddToCart
                    id="123"
                  >
                    <button
                      onClick={[Function]}
                    >
                      Add To Cart
                    </button>
                  </AddToCart>
                  <DeleteItem
                    id="123"
                  >
                    <button
                      onClick={[Function]}
                    >
                      Delete
                    </button>
                  </DeleteItem>
                </div>
              </div>
            </item_styles__Item>
          </Item>
          <Item
            data={
              Object {
                "__typename": "Item",
                "description": "Nobis culpa deleniti qui reprehenderit in.",
                "id": "456",
                "image": "foo.jpg",
                "largeImage": "large_foo.jpg",
                "price": [Function],
                "title": "Omnis reiciendis",
              }
            }
            key="456"
          >
            <item_styles__Item>
              <div
                className="item_styles__Item-sc-1wrkfgx-0 aVlSA"
              >
                <img
                  alt="Omnis reiciendis"
                  src="foo.jpg"
                />
                <Title>
                  <h3
                    className="Title-sc-1002kw6-0 eBuTHz"
                  >
                    <Link
                      href={
                        Object {
                          "pathname": "/item",
                          "query": Object {
                            "id": "456",
                          },
                        }
                      }
                    >
                      <a
                        href="/item?id=456"
                        onClick={[Function]}
                        onMouseEnter={[Function]}
                      >
                        Omnis reiciendis
                      </a>
                    </Link>
                  </h3>
                </Title>
                <PriceTag>
                  <span
                    className="PriceTag-sc-145cui2-0 cUpSNy"
                  >
                    $NaN
                  </span>
                </PriceTag>
                <p>
                  Nobis culpa deleniti qui reprehenderit in.
                </p>
                <div
                  className="buttonList"
                >
                  <Link
                    href={
                      Object {
                        "pathname": "/update",
                        "query": Object {
                          "id": "456",
                        },
                      }
                    }
                  >
                    <a
                      href="/update?id=456"
                      onClick={[Function]}
                      onMouseEnter={[Function]}
                    >
                      Edit ✏️
                    </a>
                  </Link>
                  <AddToCart
                    id="456"
                  >
                    <button
                      onClick={[Function]}
                    >
                      Add To Cart
                    </button>
                  </AddToCart>
                  <DeleteItem
                    id="456"
                  >
                    <button
                      onClick={[Function]}
                    >
                      Delete
                    </button>
                  </DeleteItem>
                </div>
              </div>
            </item_styles__Item>
          </Item>
        </div>
      </items_styles__ItemsList>
      <ItemsPagination
        currentPage={1}
      >
        <items_pagination_styles__StyledItemsPagination>
          <div
            className="items_pagination_styles__StyledItemsPagination-sc-1do2rc3-0 kTuLOm"
          >
            <Head>
              <Component
                handleStateChange={null}
                inAmpMode={false}
                reduceComponentsToState={[Function]}
              />
            </Head>
            <Link
              href={
                Object {
                  "pathname": "items",
                  "query": Object {
                    "page": 0,
                  },
                }
              }
            >
              <a
                aria-disabled={true}
                href="items?page=0"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                Prev
              </a>
            </Link>
            <p>
              Page 
              1
               of 
              2
            </p>
            <Link
              href={
                Object {
                  "pathname": "items",
                  "query": Object {
                    "page": 2,
                  },
                }
              }
            >
              <a
                aria-disabled={false}
                href="items?page=2"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                Next
              </a>
            </Link>
          </div>
        </items_pagination_styles__StyledItemsPagination>
      </ItemsPagination>
    </div>
  </items_styles__Center>
</Items>
`;

exports[`<Items /> renders loading state correctly 1`] = `
<Items
  currentPage={1}
>
  <p>
    Loading...
  </p>
</Items>
`;
